In [41]:
import warnings
warnings.filterwarnings("ignore")

import pandas as pd
import numpy as np

#Data Visualization
import seaborn as sns
import matplotlib.pyplot as plt


#Modeling

from sklearn.model_selection import train_test_split
from sklearn.preprocessing import LabelEncoder
from sklearn.preprocessing import OneHotEncoder
from sklearn.model_selection import train_test_split


from sklearn.preprocessing import MinMaxScaler
from sklearn.neighbors import KNeighborsClassifier
from sklearn.feature_selection import SelectKBest, chi2
from sklearn.ensemble import RandomForestClassifier
from sklearn.tree import DecisionTreeClassifier
from sklearn.linear_model import LogisticRegression

#Evaluation
from sklearn.tree import export_graphviz
from sklearn.metrics import classification_report
from sklearn.metrics import confusion_matrix


from scipy import stats

#Set figure size for all plots
plt.rc("figure", figsize = (16,16))

#Set fontsize for titles
plt.rc("font", size=14)

pd.set_option('display.max_columns', None)

# Project purpose

The goal of this project is to use features captured from survey data to predict how likely individuals are to receive their H1N1 and seasonal flu vaccines.

Specifically, I will be predicting two probabilities:
- h1n1_vaccine - Whether respondent received H1N1 flu vaccine.
- seasonal_vaccine - Whether respondent received seasonal flu vaccine.
> Both are binary variables: 0 = No; 1 = Yes. Some respondents didn't get either vaccine, others got only one, and some got both. This is formulated as a multilabel (and not multiclass) problem.

# Project Plan

1. Aquisition
  * Download data into local drive
2. Prepare
  * Read in data csv using pandas
  * Chekck data types and null values
  * Fill in nulls
  * Encode appropriately
  * Scale if needed
3. Explore
4. Modeling
5. Conclusions

# Acquire

- I downloaded the data from https://www.drivendata.org/competitions/66/flu-shot-learning/data/ into the same file where this analysis is being conducted
- Now I'll turn the data into a pandas dataframe for analysis and modeling

In [2]:
# Read in the csv using pandas
df = pd.read_csv('Flu_Shot_Learning_Predict_H1N1_and_Seasonal_Flu_Vaccines_-_Training_Features.csv', index_col=0)

Take a look at the data

In [3]:
df

h1n1_concern  h1n1_knowledge  behavioral_antiviral_meds  \
respondent_id                                                            
0                       1.0             0.0                        0.0   
1                       3.0             2.0                        0.0   
2                       1.0             1.0                        0.0   
3                       1.0             1.0                        0.0   
4                       2.0             1.0                        0.0   
...                     ...             ...                        ...   
26702                   2.0             0.0                        0.0   
26703                   1.0             2.0                        0.0   
26704                   2.0             2.0                        0.0   
26705                   1.0             1.0                        0.0   
26706                   0.0             0.0                        0.0   

               behavioral_avoidance  behavioral_face_mask  \
respondent_id                                               
0                               0.0                   0.0   
1                               1.0                   0.0   
2                               1.0                   0.0   
3                               1.0                   0.0   
4                               1.0                   0.0   
...                             ...                   ...   
26702                           1.0                   0.0   
26703                           1.0                   0.0   
26704                           1.0                   1.0   
26705                           0.0                   0.0   
26706                           1.0                   0.0   

               behavioral_wash_hands  behavioral_large_gatherings  \
respondent_id                                                       
0                                0.0                          0.0   
1                                1.0                          0.0   
2                                0.0                          0.0   
3                                1.0                          1.0   
4                                1.0                          1.0   
...                              ...                          ...   
26702                            0.0                          0.0   
26703                            1.0                          0.0   
26704                            1.0                          1.0   
26705                            0.0                          0.0   
26706                            0.0                          0.0   

               behavioral_outside_home  behavioral_touch_face  \
respondent_id                                                   
0                                  1.0                    1.0   
1                                  1.0                    1.0   
2                                  0.0                    0.0   
3                                  0.0                    0.0   
4                                  0.0                    1.0   
...                                ...                    ...   
26702                              1.0                    0.0   
26703                              0.0                    0.0   
26704                              0.0                    1.0   
26705                              0.0                    NaN   
26706                              0.0                    0.0   

               doctor_recc_h1n1  doctor_recc_seasonal  chronic_med_condition  \
respondent_id                                                                  
0                           0.0                   0.0                    0.0   
1                           0.0                   0.0                    0.0   
2                           NaN                   NaN                    1.0   
3                           0.0                   1.0                    1.0   
4                           0.0   

### The initial datatframe has:
- 26,707 rows where each row is one person
- 35 columns where each column is a feature this person has indicated on their survey

Let's get some more info about the datframe. We'll use .info() to see how many non-null values we have in each column and what the data types are

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 26707 entries, 0 to 26706
Data columns (total 35 columns):
h1n1_concern                   26615 non-null float64
h1n1_knowledge                 26591 non-null float64
behavioral_antiviral_meds      26636 non-null float64
behavioral_avoidance           26499 non-null float64
behavioral_face_mask           26688 non-null float64
behavioral_wash_hands          26665 non-null float64
behavioral_large_gatherings    26620 non-null float64
behavioral_outside_home        26625 non-null float64
behavioral_touch_face          26579 non-null float64
doctor_recc_h1n1               24547 non-null float64
doctor_recc_seasonal           24547 non-null float64
chronic_med_condition          25736 non-null float64
child_under_6_months           25887 non-null float64
health_worker                  25903 non-null float64
health_insurance               14433 non-null float64
opinion_h1n1_vacc_effective    26316 non-null float64
opinion_h1n1_risk          

### Initial thoughts for cleaning/prep
1. Several columns will need to be one hot encoded or label encoded
2. Nans in several columns needs to be handled
3. May need to drop certain columns
4. All the datatypes seem to be appropriate

# Prepare

I'll start by splitting the data into train and test to avoid exploring the test data which is meant to stay unseen. Then I'll perform the same clean and prep changes to each dataframe sequentially.
> There's enough data here to split this further into train and validate sets which will help to prevent overfitting by allowing signs of overfitting to be caught before applying the model to the test data. After the appropriate cleaning and preparation are complete, I will create a validate set.

In [5]:
# use the train test split function from Sklearn and add a random seed for reproducibility
train, test = train_test_split(df, random_state=123, train_size=.80, stratify=df.species)

Now I'll check to see which columns have nans, how many there are, and explore the best ways to fill those nans

In [6]:
#Write a for loop to determine what percentage of each column are nans
def percent_nans(df):
    x = ['column','n_nans', 'percentage_nans']
    missing_data_df = pd.DataFrame(columns=x)
    columns = df.columns
    for col in columns:
        column_name = col
        missing_data = df[col].isnull().sum()
        missing_in_percentage = (df[col].isnull().sum()/df[col].shape[0])*100
        
        missing_data_df.loc[len(missing_data_df)] = [column_name, missing_data, missing_in_percentage]
    return missing_data_df.sort_values(by = 'percentage_nans')

In [7]:
percent_nans(df)

column n_nans  percentage_nans
21                    age_group      0         0.000000
23                         race      0         0.000000
24                          sex      0         0.000000
29               hhs_geo_region      0         0.000000
30                   census_msa      0         0.000000
4          behavioral_face_mask     19         0.071142
5         behavioral_wash_hands     42         0.157262
2     behavioral_antiviral_meds     71         0.265848
7       behavioral_outside_home     82         0.307036
6   behavioral_large_gatherings     87         0.325757
0                  h1n1_concern     92         0.344479
1                h1n1_knowledge    116         0.434343
8         behavioral_touch_face    128         0.479275
3          behavioral_avoidance    208         0.778822
32           household_children    249         0.932340
31             household_adults    249         0.932340
16            opinion_h1n1_risk    388         1.452803
15  opinion_h1n1_vacc_effective    391         1.464036
17  opinion_h1n1_sick_from_vacc    395         1.479013
18  opinion_seas_vacc_effective    462         1.729884
19            opinion_seas_risk    514         1.924589
20  opinion_seas_sick_from_vacc    537         2.010709
13                health_worker    804         3.010447
12         child_under_6_months    820         3.070356
11        chronic_med_condition    971         3.635751
22                    education   1407         5.268282
26               marital_status   1408         5.272026
28            employment_status   1463         5.477965
27                  rent_or_own   2042         7.645936
10         doctor_recc_seasonal   2160         8.087767
9              doctor_recc_h1n1   2160         8.087767
25               income_poverty   4423        16.561201
14             health_insurance  12274        45.957989
33          employment_industry  13330        49.912008
34        employment_occupation  13470        50.436215

- It looks like only three columns have a significant number of values missing:
  - __health_insurance__
  - __employment_industry__
  - __employment_occupation__
- Look at the most common values in each feature and decide if there is a reasonable way to fill the null values

In [8]:
#What is the most common health insurance?
df.health_insurance.value_counts()

1.0    12697
0.0     1736
Name: health_insurance, dtype: int64

In [9]:
#What is the most common employment_industry?
df.employment_industry.value_counts()

fcxhlnwr    2468
wxleyezf    1804
ldnlellj    1231
pxcmvdjn    1037
atmlpfrs     926
arjwrbjb     871
xicduogh     851
mfikgejo     614
vjjrobsf     527
rucpziij     523
xqicxuve     511
saaquncn     338
cfqqtusy     325
nduyfdeo     286
mcubkhph     275
wlfvacwt     215
dotnnunm     201
haxffmxo     148
msuufmds     124
phxvnwax      89
qnlwzans      13
Name: employment_industry, dtype: int64

In [10]:
#What is the most common employment occupation?
df.employment_occupation.value_counts()

xtkaffoo    1778
mxkfnird    1509
emcorrxb    1270
cmhcxjea    1247
xgwztkwe    1082
hfxkjkmi     766
qxajmpny     548
xqwwgdyp     485
kldqjyjy     469
uqqtjvyb     452
tfqavkke     388
ukymxvdu     372
vlluhbov     354
oijqvulv     344
ccgxvspp     341
bxpfxfdn     331
haliazsg     296
rcertsgn     276
xzmlyyjv     248
dlvbwzss     227
hodpvpew     208
dcjcmpih     148
pvmttkik      98
Name: employment_occupation, dtype: int64

It is evident that __employment_industry__ and __employment_occupation__ do not have a single most overwhelming industry or occupation that could be reasonably used to fill in the remaining half empty values. I will drop these columns and not include them in the analysis

On the other hand, __health_insurance__ has 12,697 observation recoded as having insurance of the total 14,433 observations with values recorded. It may be reasonable to simply fill in the remaining 45% missing values with the label for having insurance.

### Drop the employment_industry and employment_occupation columns from dataframes

In [11]:
train = train.drop(columns =['employment_industry', 'employment_occupation'] )
test = test.drop(columns =['employment_industry', 'employment_occupation'] )

Check to make sure the columns were dropped

In [12]:
train

h1n1_concern  h1n1_knowledge  behavioral_antiviral_meds  \
respondent_id                                                            
4029                    1.0             1.0                        0.0   
10421                   1.0             2.0                        0.0   
10449                   0.0             1.0                        0.0   
19333                   1.0             0.0                        0.0   
21324                   2.0             1.0                        0.0   
...                     ...             ...                        ...   
15377                   2.0             1.0                        0.0   
21602                   3.0             2.0                        0.0   
17730                   2.0             1.0                        0.0   
15725                   1.0             2.0                        0.0   
19966                   1.0             2.0                        0.0   

               behavioral_avoidance  behavioral_face_mask  \
respondent_id                                               
4029                            0.0                   0.0   
10421                           1.0                   0.0   
10449                           0.0                   0.0   
19333                           1.0                   0.0   
21324                           1.0                   0.0   
...                             ...                   ...   
15377                           0.0                   0.0   
21602                           1.0                   0.0   
17730                           1.0                   1.0   
15725                           1.0                   0.0   
19966                           1.0                   0.0   

               behavioral_wash_hands  behavioral_large_gatherings  \
respondent_id                                                       
4029                             0.0                          0.0   
10421                            1.0                          1.0   
10449                            1.0                          0.0   
19333                            0.0                          0.0   
21324                            1.0                          0.0   
...                              ...                          ...   
15377                            1.0                          1.0   
21602                            1.0                          1.0   
17730                            1.0                          1.0   
15725                            1.0                          0.0   
19966                            0.0                          0.0   

               behavioral_outside_home  behavioral_touch_face  \
respondent_id                                                   
4029                               0.0                    0.0   
10421                              1.0                    1.0   
10449                              0.0                    1.0   
19333                              0.0                    0.0   
21324                              0.0                    0.0   
...                                ...                    ...   
15377                              0.0                    1.0   
21602                              1.0                    1.0   
17730                              1.0                    1.0   
15725                              0.0                    1.0   
19966                              0.0                    1.0   

               doctor_recc_h1n1  doctor_recc_seasonal  chronic_med_condition  \
respondent_id                                                                  
4029                        0.0                   0.0                    0.0   
10421                       1.0                   1.0                    0.0   
10449                       0.0                   0.0                    NaN   
19333                       0.0                   1.0                    NaN   
21324                       0.0   

In [13]:
test

h1n1_concern  h1n1_knowledge  behavioral_antiviral_meds  \
respondent_id                                                            
22286                   1.0             2.0                        0.0   
16228                   1.0             2.0                        0.0   
4905                    3.0             1.0                        0.0   
8947                    0.0             2.0                        0.0   
3706                    2.0             1.0                        0.0   
...                     ...             ...                        ...   
17667                   2.0             1.0                        0.0   
22560                   1.0             1.0                        0.0   
6474                    3.0             2.0                        0.0   
26073                   3.0             2.0                        1.0   
7005                    2.0             1.0                        0.0   

               behavioral_avoidance  behavioral_face_mask  \
respondent_id                                               
22286                           0.0                   0.0   
16228                           1.0                   0.0   
4905                            NaN                   0.0   
8947                            1.0                   0.0   
3706                            1.0                   0.0   
...                             ...                   ...   
17667                           0.0                   0.0   
22560                           0.0                   0.0   
6474                            1.0                   0.0   
26073                           1.0                   1.0   
7005                            0.0                   0.0   

               behavioral_wash_hands  behavioral_large_gatherings  \
respondent_id                                                       
22286                            1.0                          0.0   
16228                            1.0                          0.0   
4905                             1.0                          0.0   
8947                             1.0                          0.0   
3706                             1.0                          1.0   
...                              ...                          ...   
17667                            1.0                          1.0   
22560                            0.0                          1.0   
6474                             1.0                          1.0   
26073                            1.0                          1.0   
7005                             1.0                          1.0   

               behavioral_outside_home  behavioral_touch_face  \
respondent_id                                                   
22286                              0.0                    0.0   
16228                              0.0                    1.0   
4905                               1.0                    1.0   
8947                               0.0                    1.0   
3706                               1.0                    1.0   
...                                ...                    ...   
17667                              0.0                    1.0   
22560                              0.0                    1.0   
6474                               1.0                    1.0   
26073                              1.0                    1.0   
7005                               1.0                    1.0   

               doctor_recc_h1n1  doctor_recc_seasonal  chronic_med_condition  \
respondent_id                                                                  
22286                       0.0                   0.0                    0.0   
16228                       0.0                   0.0                    0.0   
4905                        0.0                   1.0                    0.0   
8947                        0.0                   0.0                    0.0   
3706                        0.0   

It looks like the columns were dropped appropriately. Now let's fill in null values in the remaining columns

In [14]:
#Write a function to fill in the null values with the most common occurence
def fill_null_values(train, test):
    train = train.apply(lambda x:x.fillna(x.value_counts().index[0]))
    test = test.apply(lambda x:x.fillna(x.value_counts().index[0]))

    return train, test

In [15]:
train, test = fill_null_values(train, test)

Verify that there are no more null values in either dataframe

In [16]:
train.isna().sum()

h1n1_concern                   0
h1n1_knowledge                 0
behavioral_antiviral_meds      0
behavioral_avoidance           0
behavioral_face_mask           0
behavioral_wash_hands          0
behavioral_large_gatherings    0
behavioral_outside_home        0
behavioral_touch_face          0
doctor_recc_h1n1               0
doctor_recc_seasonal           0
chronic_med_condition          0
child_under_6_months           0
health_worker                  0
health_insurance               0
opinion_h1n1_vacc_effective    0
opinion_h1n1_risk              0
opinion_h1n1_sick_from_vacc    0
opinion_seas_vacc_effective    0
opinion_seas_risk              0
opinion_seas_sick_from_vacc    0
age_group                      0
education                      0
race                           0
sex                            0
income_poverty                 0
marital_status                 0
rent_or_own                    0
employment_status              0
hhs_geo_region                 0
census_msa

In [17]:
test.isna().sum()

h1n1_concern                   0
h1n1_knowledge                 0
behavioral_antiviral_meds      0
behavioral_avoidance           0
behavioral_face_mask           0
behavioral_wash_hands          0
behavioral_large_gatherings    0
behavioral_outside_home        0
behavioral_touch_face          0
doctor_recc_h1n1               0
doctor_recc_seasonal           0
chronic_med_condition          0
child_under_6_months           0
health_worker                  0
health_insurance               0
opinion_h1n1_vacc_effective    0
opinion_h1n1_risk              0
opinion_h1n1_sick_from_vacc    0
opinion_seas_vacc_effective    0
opinion_seas_risk              0
opinion_seas_sick_from_vacc    0
age_group                      0
education                      0
race                           0
sex                            0
income_poverty                 0
marital_status                 0
rent_or_own                    0
employment_status              0
hhs_geo_region                 0
census_msa

### Encoding

Next we need to encode the columns that have objects as values and turn them into integer representations for the purpose of the classification model to predict how likely people are to get their flu vaccines

Look at which columns need to be encoded

In [18]:
train

h1n1_concern  h1n1_knowledge  behavioral_antiviral_meds  \
respondent_id                                                            
4029                    1.0             1.0                        0.0   
10421                   1.0             2.0                        0.0   
10449                   0.0             1.0                        0.0   
19333                   1.0             0.0                        0.0   
21324                   2.0             1.0                        0.0   
...                     ...             ...                        ...   
15377                   2.0             1.0                        0.0   
21602                   3.0             2.0                        0.0   
17730                   2.0             1.0                        0.0   
15725                   1.0             2.0                        0.0   
19966                   1.0             2.0                        0.0   

               behavioral_avoidance  behavioral_face_mask  \
respondent_id                                               
4029                            0.0                   0.0   
10421                           1.0                   0.0   
10449                           0.0                   0.0   
19333                           1.0                   0.0   
21324                           1.0                   0.0   
...                             ...                   ...   
15377                           0.0                   0.0   
21602                           1.0                   0.0   
17730                           1.0                   1.0   
15725                           1.0                   0.0   
19966                           1.0                   0.0   

               behavioral_wash_hands  behavioral_large_gatherings  \
respondent_id                                                       
4029                             0.0                          0.0   
10421                            1.0                          1.0   
10449                            1.0                          0.0   
19333                            0.0                          0.0   
21324                            1.0                          0.0   
...                              ...                          ...   
15377                            1.0                          1.0   
21602                            1.0                          1.0   
17730                            1.0                          1.0   
15725                            1.0                          0.0   
19966                            0.0                          0.0   

               behavioral_outside_home  behavioral_touch_face  \
respondent_id                                                   
4029                               0.0                    0.0   
10421                              1.0                    1.0   
10449                              0.0                    1.0   
19333                              0.0                    0.0   
21324                              0.0                    0.0   
...                                ...                    ...   
15377                              0.0                    1.0   
21602                              1.0                    1.0   
17730                              1.0                    1.0   
15725                              0.0                    1.0   
19966                              0.0                    1.0   

               doctor_recc_h1n1  doctor_recc_seasonal  chronic_med_condition  \
respondent_id                                                                  
4029                        0.0                   0.0                    0.0   
10421                       1.0                   1.0                    0.0   
10449                       0.0                   0.0                    0.0   
19333                       0.0                   1.0                    0.0   
21324                       0.0   

In [19]:
# def label_encode_columns(train, test):

#     encoder = LabelEncoder()
   
#     encode_list = ['rent_or_own', 'employment_status', 'marital_status', 'sex']
    
             
#     for column in encode_list:
#         train[column] = encoder.fit_transform(train[column])
#         test[column] = encoder.transform(test[column])

#         return train, test

In [20]:
# train, test = label_encode_columns(train, test)

In [21]:
def label_encode_columns(train, test):

    encoder = LabelEncoder()
   
    train['encoded_rent_or_own'] = encoder.fit_transform(train['rent_or_own'])
    train['encoded_marital_status'] = encoder.fit_transform(train['marital_status'])
    train['encoded_sex'] = encoder.fit_transform(train['sex'])

    test['encoded_rent_or_own'] = encoder.fit_transform(test['rent_or_own'])
    test['encoded_marital_status'] = encoder.fit_transform(test['marital_status'])
    test['encoded_sex'] = encoder.fit_transform(test['sex'])
    
    return train, test

In [22]:
train, test = label_encode_columns(train, test)

In [23]:
# categorical_cols = ['rent_or_own', 'employment_status', 'marital_status', 'sex']

# categorical_cols

# train[categorical_cols] = train[categorical_cols].apply(lambda col: encoder.fit_transform(col))

In [24]:
train

h1n1_concern  h1n1_knowledge  behavioral_antiviral_meds  \
respondent_id                                                            
4029                    1.0             1.0                        0.0   
10421                   1.0             2.0                        0.0   
10449                   0.0             1.0                        0.0   
19333                   1.0             0.0                        0.0   
21324                   2.0             1.0                        0.0   
...                     ...             ...                        ...   
15377                   2.0             1.0                        0.0   
21602                   3.0             2.0                        0.0   
17730                   2.0             1.0                        0.0   
15725                   1.0             2.0                        0.0   
19966                   1.0             2.0                        0.0   

               behavioral_avoidance  behavioral_face_mask  \
respondent_id                                               
4029                            0.0                   0.0   
10421                           1.0                   0.0   
10449                           0.0                   0.0   
19333                           1.0                   0.0   
21324                           1.0                   0.0   
...                             ...                   ...   
15377                           0.0                   0.0   
21602                           1.0                   0.0   
17730                           1.0                   1.0   
15725                           1.0                   0.0   
19966                           1.0                   0.0   

               behavioral_wash_hands  behavioral_large_gatherings  \
respondent_id                                                       
4029                             0.0                          0.0   
10421                            1.0                          1.0   
10449                            1.0                          0.0   
19333                            0.0                          0.0   
21324                            1.0                          0.0   
...                              ...                          ...   
15377                            1.0                          1.0   
21602                            1.0                          1.0   
17730                            1.0                          1.0   
15725                            1.0                          0.0   
19966                            0.0                          0.0   

               behavioral_outside_home  behavioral_touch_face  \
respondent_id                                                   
4029                               0.0                    0.0   
10421                              1.0                    1.0   
10449                              0.0                    1.0   
19333                              0.0                    0.0   
21324                              0.0                    0.0   
...                                ...                    ...   
15377                              0.0                    1.0   
21602                              1.0                    1.0   
17730                              1.0                    1.0   
15725                              0.0                    1.0   
19966                              0.0                    1.0   

               doctor_recc_h1n1  doctor_recc_seasonal  chronic_med_condition  \
respondent_id                                                                  
4029                        0.0                   0.0                    0.0   
10421                       1.0                   1.0                    0.0   
10449                       0.0                   0.0                    0.0   
19333                       0.0                   1.0                    0.0   
21324                       0.0   

In [25]:
# def encode(train, test):
#     # creating instance of one-hot-encoder
#     enc = OneHotEncoder()
#     # passing bridge-types-cat column (label encoded values of bridge_types)
#     enc_df = pd.DataFrame(enc.fit_transform(train[['age_group', 'education', 'race', 'income_poverty']]).toarray())
#     # merge with main df bridge_df on key values
#     train = train.join(enc_df)

#     # passing bridge-types-cat column (label encoded values of bridge_types)
#     enc_df2 = pd.DataFrame(enc.fit_transform(test[['age_group', 'education', 'race', 'income_poverty']]).toarray())
#     # merge with main df bridge_df on key values
#     test = test.join(enc_df2)
    
#     return train, test

In [26]:
# col_name = ['age_group', 'education', 'race', 'income_poverty']

In [27]:
# train, test = encode(train, test)

Look at the dataframe to ensure all the label encoded columns were added correctly

In [28]:
train

h1n1_concern  h1n1_knowledge  behavioral_antiviral_meds  \
respondent_id                                                            
4029                    1.0             1.0                        0.0   
10421                   1.0             2.0                        0.0   
10449                   0.0             1.0                        0.0   
19333                   1.0             0.0                        0.0   
21324                   2.0             1.0                        0.0   
...                     ...             ...                        ...   
15377                   2.0             1.0                        0.0   
21602                   3.0             2.0                        0.0   
17730                   2.0             1.0                        0.0   
15725                   1.0             2.0                        0.0   
19966                   1.0             2.0                        0.0   

               behavioral_avoidance  behavioral_face_mask  \
respondent_id                                               
4029                            0.0                   0.0   
10421                           1.0                   0.0   
10449                           0.0                   0.0   
19333                           1.0                   0.0   
21324                           1.0                   0.0   
...                             ...                   ...   
15377                           0.0                   0.0   
21602                           1.0                   0.0   
17730                           1.0                   1.0   
15725                           1.0                   0.0   
19966                           1.0                   0.0   

               behavioral_wash_hands  behavioral_large_gatherings  \
respondent_id                                                       
4029                             0.0                          0.0   
10421                            1.0                          1.0   
10449                            1.0                          0.0   
19333                            0.0                          0.0   
21324                            1.0                          0.0   
...                              ...                          ...   
15377                            1.0                          1.0   
21602                            1.0                          1.0   
17730                            1.0                          1.0   
15725                            1.0                          0.0   
19966                            0.0                          0.0   

               behavioral_outside_home  behavioral_touch_face  \
respondent_id                                                   
4029                               0.0                    0.0   
10421                              1.0                    1.0   
10449                              0.0                    1.0   
19333                              0.0                    0.0   
21324                              0.0                    0.0   
...                                ...                    ...   
15377                              0.0                    1.0   
21602                              1.0                    1.0   
17730                              1.0                    1.0   
15725                              0.0                    1.0   
19966                              0.0                    1.0   

               doctor_recc_h1n1  doctor_recc_seasonal  chronic_med_condition  \
respondent_id                                                                  
4029                        0.0                   0.0                    0.0   
10421                       1.0                   1.0                    0.0   
10449                       0.0                   0.0                    0.0   
19333                       0.0                   1.0                    0.0   
21324                       0.0   

Check the origional columns against the encoded ones to be clear which labels correspond to eachother

In [29]:
housing_df = train[['rent_or_own', 'encoded_rent_or_own']]
housing_df

rent_or_own  encoded_rent_or_own
respondent_id                                 
4029                  Own                    0
10421                 Own                    0
10449                 Own                    0
19333                 Own                    0
21324                Rent                    1
...                   ...                  ...
15377                 Own                    0
21602                Rent                    1
17730                 Own                    0
15725                 Own                    0
19966                 Own                    0

[21365 rows x 2 columns]

#### Own == 0
#### Rent == 1

In [30]:
married_df = train[['marital_status', 'encoded_marital_status']]
married_df

marital_status  encoded_marital_status
respondent_id                                       
4029             Not Married                       1
10421                Married                       0
10449                Married                       0
19333                Married                       0
21324            Not Married                       1
...                      ...                     ...
15377                Married                       0
21602                Married                       0
17730                Married                       0
15725                Married                       0
19966            Not Married                       1

[21365 rows x 2 columns]

#### Married == 0
#### Not Married == 1

In [31]:
gender_df = train[['sex', 'encoded_sex']]
gender_df

sex  encoded_sex
respondent_id                     
4029           Female            0
10421          Female            0
10449          Female            0
19333          Female            0
21324            Male            1
...               ...          ...
15377          Female            0
21602          Female            0
17730            Male            1
15725          Female            0
19966          Female            0

[21365 rows x 2 columns]

#### Female == 0 
#### Male == 1

### One Hot Encode Remaining Columns

In [32]:
# Encode age_group column

# Create encoder object
encoder = OneHotEncoder()

# Fit on the age_group column of the train df
encoder.fit(train[['age_group']])

# nice columns for display
cols = ['age_group_' + c for c in encoder.categories_[0]]

# Transform the column on train and test and concatenate new df onto train and test dfs
m = encoder.transform(train[['age_group']]).todense()
train = pd.concat([
    train,
    pd.DataFrame(m, columns=cols, index=train.index)
], axis=1)

m = encoder.transform(test[['age_group']]).todense()
test = pd.concat([
    test,
    pd.DataFrame(m, columns=cols, index=test.index)
], axis=1)

Check to make sure it worked

In [33]:
train

h1n1_concern  h1n1_knowledge  behavioral_antiviral_meds  \
respondent_id                                                            
4029                    1.0             1.0                        0.0   
10421                   1.0             2.0                        0.0   
10449                   0.0             1.0                        0.0   
19333                   1.0             0.0                        0.0   
21324                   2.0             1.0                        0.0   
...                     ...             ...                        ...   
15377                   2.0             1.0                        0.0   
21602                   3.0             2.0                        0.0   
17730                   2.0             1.0                        0.0   
15725                   1.0             2.0                        0.0   
19966                   1.0             2.0                        0.0   

               behavioral_avoidance  behavioral_face_mask  \
respondent_id                                               
4029                            0.0                   0.0   
10421                           1.0                   0.0   
10449                           0.0                   0.0   
19333                           1.0                   0.0   
21324                           1.0                   0.0   
...                             ...                   ...   
15377                           0.0                   0.0   
21602                           1.0                   0.0   
17730                           1.0                   1.0   
15725                           1.0                   0.0   
19966                           1.0                   0.0   

               behavioral_wash_hands  behavioral_large_gatherings  \
respondent_id                                                       
4029                             0.0                          0.0   
10421                            1.0                          1.0   
10449                            1.0                          0.0   
19333                            0.0                          0.0   
21324                            1.0                          0.0   
...                              ...                          ...   
15377                            1.0                          1.0   
21602                            1.0                          1.0   
17730                            1.0                          1.0   
15725                            1.0                          0.0   
19966                            0.0                          0.0   

               behavioral_outside_home  behavioral_touch_face  \
respondent_id                                                   
4029                               0.0                    0.0   
10421                              1.0                    1.0   
10449                              0.0                    1.0   
19333                              0.0                    0.0   
21324                              0.0                    0.0   
...                                ...                    ...   
15377                              0.0                    1.0   
21602                              1.0                    1.0   
17730                              1.0                    1.0   
15725                              0.0                    1.0   
19966                              0.0                    1.0   

               doctor_recc_h1n1  doctor_recc_seasonal  chronic_med_condition  \
respondent_id                                                                  
4029                        0.0                   0.0                    0.0   
10421                       1.0                   1.0                    0.0   
10449                       0.0                   0.0                    0.0   
19333                       0.0                   1.0                    0.0   
21324                       0.0   

In [34]:
test

h1n1_concern  h1n1_knowledge  behavioral_antiviral_meds  \
respondent_id                                                            
22286                   1.0             2.0                        0.0   
16228                   1.0             2.0                        0.0   
4905                    3.0             1.0                        0.0   
8947                    0.0             2.0                        0.0   
3706                    2.0             1.0                        0.0   
...                     ...             ...                        ...   
17667                   2.0             1.0                        0.0   
22560                   1.0             1.0                        0.0   
6474                    3.0             2.0                        0.0   
26073                   3.0             2.0                        1.0   
7005                    2.0             1.0                        0.0   

               behavioral_avoidance  behavioral_face_mask  \
respondent_id                                               
22286                           0.0                   0.0   
16228                           1.0                   0.0   
4905                            1.0                   0.0   
8947                            1.0                   0.0   
3706                            1.0                   0.0   
...                             ...                   ...   
17667                           0.0                   0.0   
22560                           0.0                   0.0   
6474                            1.0                   0.0   
26073                           1.0                   1.0   
7005                            0.0                   0.0   

               behavioral_wash_hands  behavioral_large_gatherings  \
respondent_id                                                       
22286                            1.0                          0.0   
16228                            1.0                          0.0   
4905                             1.0                          0.0   
8947                             1.0                          0.0   
3706                             1.0                          1.0   
...                              ...                          ...   
17667                            1.0                          1.0   
22560                            0.0                          1.0   
6474                             1.0                          1.0   
26073                            1.0                          1.0   
7005                             1.0                          1.0   

               behavioral_outside_home  behavioral_touch_face  \
respondent_id                                                   
22286                              0.0                    0.0   
16228                              0.0                    1.0   
4905                               1.0                    1.0   
8947                               0.0                    1.0   
3706                               1.0                    1.0   
...                                ...                    ...   
17667                              0.0                    1.0   
22560                              0.0                    1.0   
6474                               1.0                    1.0   
26073                              1.0                    1.0   
7005                               1.0                    1.0   

               doctor_recc_h1n1  doctor_recc_seasonal  chronic_med_condition  \
respondent_id                                                                  
22286                       0.0                   0.0                    0.0   
16228                       0.0                   0.0                    0.0   
4905                        0.0                   1.0                    0.0   
8947                        0.0                   0.0                    0.0   
3706                        0.0   

Excellent! It worked the way I wanted. Now I have a column with a 1 if the observation falls into that category and a zero if it does not. 

Now repeat for the remaining columns

In [35]:
# Encode education column

# Create encoder object
encoder = OneHotEncoder()

# Fit on the age_group column of the train df
encoder.fit(train[['education']])

# nice columns for display
cols = ['education_' + c for c in encoder.categories_[0]]

# Transform the column on train and test and concatenate new df onto train and test dfs
m = encoder.transform(train[['education']]).todense()
train = pd.concat([
    train,
    pd.DataFrame(m, columns=cols, index=train.index)
], axis=1)

m = encoder.transform(test[['education']]).todense()
test = pd.concat([
    test,
    pd.DataFrame(m, columns=cols, index=test.index)
], axis=1)

In [36]:
# Encode race column

# Create encoder object
encoder = OneHotEncoder()

# Fit on the age_group column of the train df
encoder.fit(train[['race']])

# nice columns for display
cols = ['race_' + c for c in encoder.categories_[0]]

# Transform the column on train and test and concatenate new df onto train and test dfs
m = encoder.transform(train[['race']]).todense()
train = pd.concat([
    train,
    pd.DataFrame(m, columns=cols, index=train.index)
], axis=1)

m = encoder.transform(test[['race']]).todense()
test = pd.concat([
    test,
    pd.DataFrame(m, columns=cols, index=test.index)
], axis=1)

In [37]:
# Encode income_poverty column

# Create encoder object
encoder = OneHotEncoder()

# Fit on the age_group column of the train df
encoder.fit(train[['income_poverty']])

# nice columns for display
cols = ['income_poverty_' + c for c in encoder.categories_[0]]

# Transform the column on train and test and concatenate new df onto train and test dfs
m = encoder.transform(train[['income_poverty']]).todense()
train = pd.concat([
    train,
    pd.DataFrame(m, columns=cols, index=train.index)
], axis=1)

m = encoder.transform(test[['income_poverty']]).todense()
test = pd.concat([
    test,
    pd.DataFrame(m, columns=cols, index=test.index)
], axis=1)

In [38]:
train

h1n1_concern  h1n1_knowledge  behavioral_antiviral_meds  \
respondent_id                                                            
4029                    1.0             1.0                        0.0   
10421                   1.0             2.0                        0.0   
10449                   0.0             1.0                        0.0   
19333                   1.0             0.0                        0.0   
21324                   2.0             1.0                        0.0   
...                     ...             ...                        ...   
15377                   2.0             1.0                        0.0   
21602                   3.0             2.0                        0.0   
17730                   2.0             1.0                        0.0   
15725                   1.0             2.0                        0.0   
19966                   1.0             2.0                        0.0   

               behavioral_avoidance  behavioral_face_mask  \
respondent_id                                               
4029                            0.0                   0.0   
10421                           1.0                   0.0   
10449                           0.0                   0.0   
19333                           1.0                   0.0   
21324                           1.0                   0.0   
...                             ...                   ...   
15377                           0.0                   0.0   
21602                           1.0                   0.0   
17730                           1.0                   1.0   
15725                           1.0                   0.0   
19966                           1.0                   0.0   

               behavioral_wash_hands  behavioral_large_gatherings  \
respondent_id                                                       
4029                             0.0                          0.0   
10421                            1.0                          1.0   
10449                            1.0                          0.0   
19333                            0.0                          0.0   
21324                            1.0                          0.0   
...                              ...                          ...   
15377                            1.0                          1.0   
21602                            1.0                          1.0   
17730                            1.0                          1.0   
15725                            1.0                          0.0   
19966                            0.0                          0.0   

               behavioral_outside_home  behavioral_touch_face  \
respondent_id                                                   
4029                               0.0                    0.0   
10421                              1.0                    1.0   
10449                              0.0                    1.0   
19333                              0.0                    0.0   
21324                              0.0                    0.0   
...                                ...                    ...   
15377                              0.0                    1.0   
21602                              1.0                    1.0   
17730                              1.0                    1.0   
15725                              0.0                    1.0   
19966                              0.0                    1.0   

               doctor_recc_h1n1  doctor_recc_seasonal  chronic_med_condition  \
respondent_id                                                                  
4029                        0.0                   0.0                    0.0   
10421                       1.0                   1.0                    0.0   
10449                       0.0                   0.0                    0.0   
19333                       0.0                   1.0                    0.0   
21324                       0.0   

In [39]:
test

h1n1_concern  h1n1_knowledge  behavioral_antiviral_meds  \
respondent_id                                                            
22286                   1.0             2.0                        0.0   
16228                   1.0             2.0                        0.0   
4905                    3.0             1.0                        0.0   
8947                    0.0             2.0                        0.0   
3706                    2.0             1.0                        0.0   
...                     ...             ...                        ...   
17667                   2.0             1.0                        0.0   
22560                   1.0             1.0                        0.0   
6474                    3.0             2.0                        0.0   
26073                   3.0             2.0                        1.0   
7005                    2.0             1.0                        0.0   

               behavioral_avoidance  behavioral_face_mask  \
respondent_id                                               
22286                           0.0                   0.0   
16228                           1.0                   0.0   
4905                            1.0                   0.0   
8947                            1.0                   0.0   
3706                            1.0                   0.0   
...                             ...                   ...   
17667                           0.0                   0.0   
22560                           0.0                   0.0   
6474                            1.0                   0.0   
26073                           1.0                   1.0   
7005                            0.0                   0.0   

               behavioral_wash_hands  behavioral_large_gatherings  \
respondent_id                                                       
22286                            1.0                          0.0   
16228                            1.0                          0.0   
4905                             1.0                          0.0   
8947                             1.0                          0.0   
3706                             1.0                          1.0   
...                              ...                          ...   
17667                            1.0                          1.0   
22560                            0.0                          1.0   
6474                             1.0                          1.0   
26073                            1.0                          1.0   
7005                             1.0                          1.0   

               behavioral_outside_home  behavioral_touch_face  \
respondent_id                                                   
22286                              0.0                    0.0   
16228                              0.0                    1.0   
4905                               1.0                    1.0   
8947                               0.0                    1.0   
3706                               1.0                    1.0   
...                                ...                    ...   
17667                              0.0                    1.0   
22560                              0.0                    1.0   
6474                               1.0                    1.0   
26073                              1.0                    1.0   
7005                               1.0                    1.0   

               doctor_recc_h1n1  doctor_recc_seasonal  chronic_med_condition  \
respondent_id                                                                  
22286                       0.0                   0.0                    0.0   
16228                       0.0                   0.0                    0.0   
4905                        0.0                   1.0                    0.0   
8947                        0.0                   0.0                    0.0   
3706                        0.0   

Turn these into functions and add to prepare.py. I'll just do one example here to test if it works then comment it out to make sure I don't have duplicate columns moving forward. The other functions will only be added to the .py file.

In [40]:
# def ohe_income_poverty(train,test):
#     # Encode income_poverty column

#     # Create encoder object
#     encoder = OneHotEncoder()

#     # Fit on the age_group column of the train df
#     encoder.fit(train[['income_poverty']])

#     # nice columns for display
#     cols = ['income_poverty_' + c for c in encoder.categories_[0]]

#     # Transform the column on train and test and concatenate new df onto train and test dfs
#     m = encoder.transform(train[['income_poverty']]).todense()
#     train = pd.concat([
#         train,
#         pd.DataFrame(m, columns=cols, index=train.index)
#     ], axis=1)

#     m = encoder.transform(test[['income_poverty']]).todense()
#     test = pd.concat([
#         test,
#         pd.DataFrame(m, columns=cols, index=test.index)
#     ], axis=1)

#     return train, test

### Scaling

- The following columns are either opions on a scale of 1-5 or a count of number of children or adults in a household:
  - h1n1_concern
  - h1n1_knowledge
  - opinion_h1n1_vacc_effective
  - opinion_h1n1_risk
  - opinion_h1n1_sick_from_vacc
  - opinion_seas_vacc_effective
  - opinion_seas_risk
  - opinion_seas_sick_from_vac
  - household_adults
  - household_children
- All other features are on a scale of 0-1. I will apply a MinMax Scaler to the above columns to get them also on a 0-1 scale to avoid weighting issues in the models to come. 